My very first pull request

aliyss opened the pull request:
I came over your repo and since I found it particularly upto date I was wondering if you mind me adding some stuff or not. Here's a list of changes I made:

Added: isOnline and kind
I added the isOnline and the kind. Although kind only returns "chat" and "group" it might be good in future to check for enterprise-chats or similar in case it comes out. In my opinion isOnline can be very helpful.

Removed: fromMe check
Adding to that I removed the fromMe as I believe this should be an if else provided by the user directly.

Added: Return of the message, currently it only returns the string success, but i plan on making it work so that it returns the message sent.

aliyss committed:
added presence and kind

aliyss committed:
added isOnline and kind

aliyss committed:
removed console.log

aliyss committed:
will return last message

Aliyss commented:
as promised, i added a return. After sending the message will be retrieved and returned to the user.

basically allows the user to use the following format:

let m = await client.sendMessage(message.from, content);

pedroslopez commented on src/util/Injected.js:
```diff
@@ -176,21 +176,27 @@ exports.LoadCustomSerializers = () => {
             res.groupMetadata = chat.groupMetadata.serialize();
         }
 
+        if (!res.isGroup) {
```
This `if` here isn't really needed since all chat objects have a presence, even though it will always be `false` for groups.

pedroslopez commented on src/structures/Chat.js:
```diff
@@ -17,6 +17,8 @@ class Chat extends Base {
         this.id = data.id;
 
         this.name = data.formattedTitle;
+        this.isOnline = data.isOnline;
```
Since this only really applies for private chats, consider patching this in the `PrivateChat` model, instead of the more general `Chat` one.

pedroslopez commented on src/Client.js:
```diff
@@ -137,9 +137,12 @@ class Client extends EventEmitter {
-        }, chatId, message)
+        let last_message = await this.pupPage.evaluate(async (chatId, message) => {
+            await Store.SendMessage(Store.Chat.get(chatId), message);
+            return Store.Chat.get(chatId).msgs._last.serialize()
```
This is actually a pretty nice idea, however, this implementation is error prone since you don't have any guarantees that the last message in the chat is the message you sent. 

I was able to break this with the example in the repo by putting my other phone in airplane mode, sending `!ping` along with a couple more messages, and then turning airplane mode off. All the messages arrive to our client seemingly at the same time, but in between the time it takes to send a reply to the `!ping` command and when you save that `last_message`, another one of the messages that were coming in was received, so the message sent by the client is in fact *not* the last message in the chat.

Aliyss replied:
Thanks for the fast reply. Will work on the other stuff after the weekend and plan on adding a lot more. Any idea on how it would be possible to successfully retrieve the sent message?because returning Store.SendMessage only outputs the string "success".

Aliyss replied:
@pedroslopez Pinging you in case you didn't see it. Btw. do you use discord by any chance?

pedroslopez replied:
Hey! Hadn't seen this. I'm not sure to be honest how we would be able to reliably get the latest message. Maybe it can be done matching the last message based on the content itself and things like user? Not sure.

Yeah I'm on discord, nacho#6294 if you want to message me.

pedroslopez replied:
I'm still not convinced on this issue... maybe leave this out until we find a more definitive way to get the sent message?

pedroslopez reviewed (changes requested):
Hey there! Thanks a lot for contributing! I really appreciate it. 

I've left you some comments you should take a look at before accepting this PR, but overall really nice ideas and additions.

In the future please remember to make the PR titles more descriptive, indicating a bit about what's contained inside. 😄

aliyss committed:
Added the changes. Ready for review.

- Client now only returns message if content is the same.
- removed if from isOnline
- PrivateChat now has isOnline instead of the Chat itself (don't know if this has been implemented correctly)

aliyss committed:
Everything Works:

- books added

aliyss committed:
fixed spacing

aliyss committed:
fixed spacing again

aliyss committed:
Added a mentions list to see what users were mentioned in the message.

aliyss committed:
Added Message mentioning:

To mention a user pass an object into the message with the following format:

message = {
 body: "Your text @1234567"
 mentions: ['1234567@c.us']
}

Without mentions use normal string.

aliyss committed:
fixed minimal stuff so that it theoretically works.

aliyss committed:
nvm before... made some fixes... Now it really works.

aliyss committed:
Everything Works: Removed the spacing check.

aliyss committed:
Everything Works: spacing fixed.

pedroslopez commented on src/Client.js:
```diff
@@ -134,12 +135,37 @@ class Client extends EventEmitter {
+        let last_message = await this.pupPage.evaluate(async (chatId, msg) => {
+            msg.mentionedJidList = [];
+            for (let i = 0; i < msg.mentions.length; i++) {
+                let user = await Store.Contact.serialize().find(x => x.id.user === msg.mentions[i].split("@")[0]);
```
Since you're using Contact, would this work if the user isn't one of your contacts / if you don't have a chat open with them?

Also, maybe we should just pass an extra argument to the sendMessage function that's an object with these other properties, leaving the first argument as a string as is

pedroslopez reviewed (changes requested):
Sorry it took so long to review this... I got insanely busy. Anyway, we're close to merging :)

Kouthk commented:
@Aliyss This function would be extremely useful to me, do you still think about finishing it? say yes pls haha

Aliyss commented:
@Kouthk afaik I integrated it and split up the pull requests. And as I didn't know how to use pull requests before this one contains a lot of diverse stuff.

If you can tell me what you need I can check if it has been added.

Kouthk commented:
I would like the function to display the last message sent by my device, something like "Last message from me". 
I developed a function for this, but it is not very beautiful or performative, as I am a beginner in the universe of programming...
I don't want to use the one developed by me outside the test environment

Aliyss commented:
@Kouthk You can check out the code below to see how to save your last message into a variable.


```js
let last_message;

client.on('message_create', (msg) => {
  // Fired on all message creations, including your own
  if (msg.fromMe) {
      last_message = msg
  }
});
```

or 

client.sendMessage(number, message);

```js
let last_message;

client.on('message', async msg => {
    console.log('MESSAGE RECEIVED', msg);

    if (msg.body === '!ping reply') {
        // Send a new message as a reply to the current one
        last_message = await msg.reply('pong');
    }
})
```

If you need help with your code you can try asking in the [discord server](https://discord.gg/wyKybbF).
